### PR TITLE
fix: add concurrency example, add semaphore builtin functions

### PR DESCRIPTION
**Pull request by wxiaoyun**
### Description
- Add `sem_create` as a builtin function
- Add `sem_set` builtin function
- Add concurrency examples
- Fix clippy warnings

### CODE REVIEW

File: compiler/oxidate/src/compiler.rs [hunk 1]
```diff
@@ -478,7 +478,6 @@ mod tests {
         assert_eq!(res, exp);
     }
 
-    use bytecode::Value::*;
```
Comment (wxiaoyun) [inline compiler/oxidate/src/compiler.rs:481]: Some clippy warning for unused

File: example/concurrency-02.rst [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+let tid_1 = spawn increment(l00);
+let tid_2 = spawn increment(100);
+let tid_3 = spawn increment(100);
+
+join tid_3;
+join tid_2;
+join tid_1;
+
+count
```
Comment (wxiaoyun) [inline example/concurrency-02.rst]: Example code to show race condition

File: example/concurrency-03.rst [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+let tid_1 = spawn increment(l00);
+let tid_2 = spawn increment(100);
+let tid_3 = spawn increment(100);
+
+join tid_3;
+join tid_2;
+join tid_1;
+
+count
```
Comment (wxiaoyun) [inline example/concurrency-03.rst]: And how semaphore prevents race condition

File: example/concurrency-03.rst [hunk 4]
```diff
@@ -0,0 +1,19 @@
+let count = 0;
+
+fn increment(times: int) {
+  let i = 0;
+  while i < times {
```
Comment (leonidas1712) [inline example/concurrency-03.rst:5]: Syntax wise, for loops I think I will just use the `loop` keyword with optional condition i.e `loop i < times { .. }` to stop after just like how Go has `for` and `for <cond>`.

**Review by leonidas1712 — APPROVED** [review]
LGTM, examples can be finalized later once I add frontend. Thanks

**Pull request merged**